src/ray/raylet/worker_pool.h

```diff
@@ -61,10 +61,15 @@ enum PopWorkerStatus {
   RuntimeEnvCreationFailed = 4,
 };
 
-/// \Return true if the worker was used. Otherwise, return false and the worker will be
-/// returned to the worker pool.
+/// \param[in] worker The started worker instance. Nullptr if worker is not started.
+/// \param[in] status The pop worker status. OK if things go well. Otherwise, it will
+/// contain the error status. \param[in] runtime_env_setup_error_message The error message
+/// when runtime env setup is failed. This should be empty unless status ==
+/// RuntimeEnvCreationFailed. \return true if the worker was used. Otherwise, return false
+/// and the worker will be returned to the worker pool.
 using PopWorkerCallback = std::function<bool(
-    const std::shared_ptr<WorkerInterface> worker, PopWorkerStatus status)>;
+    const std::shared_ptr<WorkerInterface> worker, PopWorkerStatus status,
+    const std::string &runtime_env_setup_error_message)>;
```

Comment:
It's a little ugly that we have a specific parameter being passed around just for runtime_env errors -- is it possible to instead have a generic "message" field on `PopWorkerStatus` that can we reused for other statuses? That would be preferable.

Reply:
I agree, and actually this is what I tried in the first iteration.

The awkwardness of "general message" is that we don't need them except runtime env failure. For other status, the message will be 
- unused. 
- not useful (It will just contain duplicated information from PopWorkerStatus).
- Or it's going to be empty.

Imo, having such code path is more confusing, so I went to this direction. Agreed it is still ugly...

Actually, this is the reason why I think the runtime env logic should move to the cluster task manager. We can remove runtime-env specific code here.

So, I think there are 2 options here;
- Keep it as it is and do refactoring in a couple months.
- Just use general context field, and make them empty or add unuseful payload.

What's your preference?

Reply:
Agreed that adding a generic path that's mostly "dead code" is not ideal. If we plan to clean this codepath up anyways, I'm fine w/ leaving it as-is for now.